src/hotspot/share/memory/metaspace.hpp

```diff
@@ -159,6 +159,10 @@ class Metaspace : public AllStatic {
     return NOT_LP64(false) LP64_ONLY(UseCompressedClassPointers);
   }
 
+  static bool is_compressed_klass_ptr(const Klass* k) {
+    return using_class_space() && (is_in_class_space(k) || is_in_shared_metaspace(k));
+  }
+
```

Comment:
I propose to drop this, and instead add a utility function to `CompressedKlassPointers` like this:

```
// Returns true if p falls into the narrow Klass encoding range
inline bool CompressedKlassPointers::is_in_encoding_range(const void* p) {
  return _base != nullptr && p >= _base && p < (_base + _range);
}
```
(Probably  the `_base != nullptr` could even be left out, since `_range==0` and `_base==nullptr` for -UseCompressedClassPointers)

And then use that function in `jfrTraceIdKlass.cpp`. That file needs to use `CompressedKlassPointers` anyway because it needs to encode the Klass*.

This avoids having to rely on the exact composition of the memory regions inside the encoding range. What counts is whether the Klass pointer points into the narrow Klass encoding range.

Essentially, with CDS, memory looks like this:

```
encoding                                                               encoding              
base                                                                   end
|-----------------------------------------------------------------------|
|----CDS---| |--------------------class space---------------------------|
```

Reply:
oh yes that's much better.